apps/api/src/persistence/repositories/domain-data-exporter.ts

```diff
@@ -10,20 +9,57 @@ import { InternalError, isInternalError } from '../../lib/errors/InternalError';
 import { ResultOrError } from '../../types/ResultOrError';
 import { REPOSITORY_PROVIDER_TOKEN } from '../constants/persistenceConstants';
 
+type InMemoryDomainSnapshotIncludingErrors = Record<AggregateType, ResultOrError<Aggregate>[]>;
+
 export class DomainDataExporter {
     constructor(
         @Inject(REPOSITORY_PROVIDER_TOKEN) private readonly repositoryProvider: IRepositoryProvider
     ) {}
 
     // TODO consider exposing a filter callback here
     async fetchSnapshot(): Promise<DeluxeInMemoryStore> {
-        const categoryTree = await this.repositoryProvider.getCategoryRepository().fetchTree();
+        const snapshotWithErrors = await this.fetchAllAggregateRoots();
+
+        return Object.entries(snapshotWithErrors)
+            .filter(
+                (
+                    aggregateTypeAndInstances
+                ): aggregateTypeAndInstances is [AggregateType, Aggregate[]] => {
+                    const [_, instances] = aggregateTypeAndInstances;
+
+                    if (instances.some(isInternalError)) {
+                        throw new InternalError(
+                            `invalid data encountered in snapshot: ${instances}`
+                        );
+                    }
 
-        if (categoryTree.some(isInternalError)) {
-            throw new InternalError(
-                `Cannot export snapshot due to invalid category tree: ${categoryTree}`
+                    return true;
+                }
+            )
+            .reduce(
+                (snapshot, [aggregateType, instances]) =>
+                    snapshot.appendAggregates(aggregateType, instances),
+                new DeluxeInMemoryStore()
             );
-        }
+    }
+
+    async validateAllInvariants(): Promise<InternalError[]> {
+        const snapshotWithErrors = await this.fetchAllAggregateRoots();
+
+        const allErrors = Object.entries(snapshotWithErrors).reduce(
+            (allErrors: InternalError[], [_, instancesOrErrors]) => {
```

Comment:
Why not use filter here?

Reply:
We use filter on the array-valued properties one-by-one, hence the reduce. Better add a comment since I couldn't easily ready it myself on review.